### ISSUE TITLE
Circuit switch gets stuck when dragged too quickly

### ISSUE THREAD

**Issue by arnabp** [phetsims/capacitor-lab-basics#195]
On Capacitance screen it only seems to get stuck on the solder balls. On Light Bulb screen, switch can get stuck in any position depending on how quickly initial dragging speed is.

![capacitor lab drag bug 1](https://user-images.githubusercontent.com/8583364/31967365-ef640a4e-b8ca-11e7-82ae-551aa9f9bcf1.gif)

![capacitor lab drag bug 2](https://user-images.githubusercontent.com/8583364/31967372-f4fe087e-b8ca-11e7-829a-1bc7bbd64fe0.gif)

For phetsims/QA#56

Troubleshooting information (do not edit):
Name: ‪Capacitor Lab: Basics‬
URL: https://www.colorado.edu/physics/phet/dev/html/capacitor-lab-basics/1.5.0-dev.1/capacitor-lab-basics_en.html
Version: 1.5.0-dev.1 2017-10-18 16:42:15 UTC
Features missing: touch
User Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/61.0.3163.100 Safari/537.36
Language: en-US
Window: 1920x1006
Pixel Ratio: 1/1
WebGL: disabled (Web page was not allowed to create a WebGL context. Web page was not allowed to create a WebGL context. )
Dependencies JSON: {"assert":{"sha":"928741cf","branch":"master"},"axon":{"sha":"8a8ff7d2","branch":"master"},"babel":{"sha":"0b5a4452","branch":"master"},"brand":{"sha":"cfca902d","branch":"master"},"capacitor-lab-basics":{"sha":"5fdf1748","branch":"master"},"chipper":{"sha":"b8a3c09a","branch":"master"},"dot":{"sha":"84bc5146","branch":"master"},"joist":{"sha":"e5c2e673","branch":"master"},"kite":{"sha":"067109a8","branch":"master"},"phet-core":{"sha":"054b2d6d","branch":"master"},"phetcommon":{"sha":"b13138d2","branch":"master"},"query-string-machine":{"sha":"c74e454e","branch":"master"},"scenery":{"sha":"c9c3052a","branch":"master"},"scenery-phet":{"sha":"124c1c25","branch":"master"},"sherpa":{"sha":"0d590744","branch":"master"},"sun":{"sha":"082fee37","branch":"master"},"tandem":{"sha":"98e9917c","branch":"master"},"twixt":{"sha":"a2eaa10a","branch":"master"}}

**Comment by lmulhall-phet** [comment]
I haven't seen this exact issue, but I have noticed that if you have the arm of wire selected and you try to drag it from one junction to another, it won't move if you drag your cursor below the joint. Not sure if this is intended behavior.

![drawing](https://user-images.githubusercontent.com/31285590/32068483-b5e994ee-ba43-11e7-871a-999198776b72.jpeg)

**Comment by jonathanolson** [comment]
Some of this seems somewhat intentional:
```js
// make sure that the switch does not snap to connection points if the user drags beyond limiting angles
if ( Math.abs( angleDifference( currentAngle, lastAngle ) ) >= Math.PI / 4 ) {
```

@arouinfar, what is the desired behavior here?

**Comment by arouinfar** [comment]
For the life of me I cannot reproduce the behavior described by @arnabp and @lmulhall-phet. That said, I cannot think of a productive or pedagogically useful reason to purposefully prevent the switch from moving in these situations @jonathanolson.

**Comment by jonathanolson** [comment]
Addressed with https://github.com/phetsims/capacitor-lab-basics/commit/d5695f2cd451b4348c59e3028154c292302a1866.

@arouinfar, can you try out the behavior (notably dragging "under", like in https://github.com/lmulhall-phet)?

**Comment by arouinfar** [comment]
@jonathanolson I could not reliably reproduce the original issue, so I think this is best verified by @arnabp, @lmulhall-phet or @phet-steele.

**Comment by jonathanolson** [comment]
This was:

1. A change that fixes the bug, since it places no restrictions on how far the switch can be moved in one time-step.
2. A change to how the switch behaves when the mouse is dragged below the (top) switch's hinge. It used to not move at all, and now it will go to the closest option.

Since (2) is a behavioral change, I wanted to ensure that you liked the behavior.

I agree (1) should be reviewed by someone who can reproduce.

**Comment by arouinfar** [comment]
Thanks for clarifying @jonathanolson! When playing with master, the switch behavior still felt very natural to me.

**Comment by phet-steele** [comment]
Firstly, original issue is fixed.
Secondly, a slack conversation with @jonathanolson.

SD
1. Why can't dragging under the switch be continuous, instead of causing the switch to jump between min/max angles? Are the angles just too small at that point?

    1a. Since only min/max angles are being reached, you can't end in the middle position when 
    dragging under the switch (2nd screen).
2.  The further below the switch my mouse is, why do I have to drag so far to the right to make the switch move (1st screen)?

JO:
(1) totally possible, but the code was initially written thinking about the "angle relative to the hinge point". Feel free to propose behavioral changes and I can implement it.

(1a) That's inconvenient, sounds like a good reason to change it

(2) Imagine a vertical line below the hinge point (where if the mouse is on one side, it will be in the left configuration, and otherwise on the right configuration). Is that happening, or is something else funky going on?
Basically... totally happy to proactively change it to better behavior, let me know

SD:
the “vertical line” feels like it might be rotated 45 degrees or so [negative slope]
so I have to move farther right to cross that line
the further down I go

JO:
Ahh, in the 3-option switch it is vertical
In the other one, it is snapping to the closest angle, so it's a line that is "between" the "closed" and "open" directions
(end)

Video showing (2).
![clb09](https://user-images.githubusercontent.com/5863899/32187894-e7ec32be-bd6b-11e7-9a41-e4fba464a531.gif)

@arouinfar I don't think these are critical (except 1a?), but they do leave a bad taste in my mouth. I think there is room for improvement, but up to you.

**Comment by arouinfar** [comment]
> 1. Why can't dragging under the switch be continuous, instead of causing the switch to jump between min/max angles? Are the angles just too small at that point?

Continuous dragging makes sense to me. I don't think the drag behavior was ever specified, so my guess is that it was a dev choice. @jonathanolson I would prefer that the switch move continuously when dragged, regardless of whether the cursor/finger is above or below the switch.

> 1a. Since only min/max angles are being reached, you can't end in the middle position when
dragging under the switch (2nd screen).

Great catch @phet-steele. This seems super problematic to me, so please fix it @jonathanolson.

> 2. The further below the switch my mouse is, why do I have to drag so far to the right to make the switch move (1st screen)?

Also seems odd to me. The only alternative I can think of would be to piecewise define the imaginary line (see below). If dragging above the top hinge, the line would remain as-is. If dragging below the hinge, you would need to cross vertical switch position. (We would do something similar for the bottom switch too.)  Not sure how feasible/maintainable this would be, so open to suggestions from @jonathanolson or @phet-steele.
![image](https://user-images.githubusercontent.com/8419308/32190646-0c853622-bd74-11e7-8b4f-88763468e063.png)

**Comment by jonathanolson** [comment]
Can @arouinfar and @phet-steele check the behavior on master to see if that is good?

**Comment by arouinfar** [comment]
I really like the new behavior @jonathanolson!! I'll leave the issue open for @phet-steele to review too.
